File: src/content/1.8/code/haskell/snippet29.hs
class Profunctor p where
    dimap :: (a -> b) -> (c -> d) -> p b c -> p a d
    dimap f g = lmap f . rmap g
    lmap :: (a -> b) -> p b c -> p a c
    lmap f = dimap f id
    rmap :: (b -> c) -> p a b -> p a c
    rmap = dimap id

File: src/content/1.8/code/scala/snippet29.scala
trait Profunctor[F[_, _]] {
  def dimap[A, B, C, D]: (A => B) => (C => D) => F[B, C] => F[A, D] = f => g =>
    lmap(f) compose rmap[B, C, D](g)

  def lmap[A, B, C]: (A => B) => F[B, C] => F[A, C] = f =>
    dimap(f)(identity[C])

  def rmap[A, B, C]: (B => C) => F[A, B] => F[A, C] =
    dimap[A, A, B, C](identity[A])
}
